File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/utils/hardware/physical/request/GyroRequest.java
package org.firstinspires.ftc.teamcode.utils.hardware.physical.request;

import com.michaell.looping.ScriptParameters;
import com.qualcomm.robotcore.hardware.GyroSensor;
import com.qualcomm.robotcore.hardware.HardwareMap;
import org.firstinspires.ftc.teamcode.utils.hardware.physical.inputs.GyroData;
import org.firstinspires.ftc.teamcode.utils.hardware.physical.inputs.GyroData;
import org.firstinspires.ftc.teamcode.utils.hardware.physical.inputs.GyroData;

public class GyroRequest extends ScriptParameters.Request {

    private final GyroSensor SENSOR;

    public GyroRequest(String name, HardwareMap hardwareMap) {
        super(name);
        SENSOR = hardwareMap.get(GyroSensor.class, name);
        SENSOR.resetDeviceConfigurationForOpMode();
        try {
            SENSOR.calibrate();
            while(SENSOR.isCalibrating());
        } catch(UnsupportedOperationException ignored) {}
    }

    @Override
    public Object issueRequest(Object o) {
        return new GyroData(SENSOR.getHeading(), new int[] { SENSOR.rawX(), SENSOR.rawY(), SENSOR.rawZ() },
            SENSOR.status());
    }

    @Override
    public Class getOutputType() {
        return GyroData.class;
    }

    @Override
    public Class getInputType() {
        return Object.class;
    }

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/utils/hardware/requests/GyroRequest.kt
package org.firstinspires.ftc.teamcode.utils.hardware.requests

import com.michaell.looping.ScriptParameters
import com.qualcomm.robotcore.hardware.GyroSensor
import com.qualcomm.robotcore.hardware.HardwareMap
import org.firstinspires.ftc.teamcode.utils.hardware.data.GyroData

class GyroRequest(name: String, hardwareMap: HardwareMap) : ScriptParameters.Request(name) {
    val gyro: GyroSensor

    init {
        gyro = hardwareMap.get(GyroSensor::class.java, name)
        gyro.resetDeviceConfigurationForOpMode()
        try {
            gyro.calibrate()
            while (gyro.isCalibrating);
        } catch (ignored: UnsupportedOperationException) {
        }
    }

    override fun issueRequest(o: Any): Any {
        return GyroData(
            gyro.heading, intArrayOf(gyro.rawX(), gyro.rawY(), gyro.rawZ()),
            gyro.status()
        )
    }

    override fun getOutputType(): Class<*> {
        return GyroData::class.java
    }

    override fun getInputType(): Class<*> {
        return Any::class.java
    }
}